zoo/zoo.py

```diff
@@ -0,0 +1,281 @@
+        """Zoo animals have names, species, cages and are created 'Alive'."""
+        self.name = name
+        self.status = self.STATUS_ALIVE
+        self.species = None
+        self.animal_type = None
+        self.sit_rep = None
+        self.competition_score = None
+        self.meat_quality = None
+        if cage:
+            self.cage = cage
+
+    def __str__(self):
+        """Human readable string returned when the class object is printed."""
+        return self.name
+
+    def __repr__(self):
+        """Class representation, returned when class object is called."""
+        return '<%s: %s (%s)>' % (self.__class__.__name__, self, self.species)
+
+    def get_status(self):
+        """Return a human readable status code"""
+        return ANIMAL_STATUS[self.status]
+
+    def get_type(self):
+        """Return a human readable animal type"""
+        return ANIMAL_TYPE[self.animal_type]
+
+    def get_quality(self):
+        """Return a human readable animal type"""
+        return MEAT_QUALITY[self.meat_quality]
+
+    def get_competition_score(self):
+        """Identify the correct relationships for a given animal.
+
+        Relationships are calculated based on the TYPE of animal and the quality
+        of it's meat.
+
+        For example:
+            LIONS are carnivores and eat only TENDER meat.
+
+        Current scores:
+            LION = 1
+            HYENA = 2
+            WILDEBEEST = 3
+            GAZELLE = 6
+
+        Lower scores indicate that animals are more likely to eat and less likely to be eaten.
+        """
+        type_score = next(i for i, v in enumerate(ANIMAL_TYPES) if v == self.animal_type)
+        quality_score = next(i for i, v in enumerate(ANIMAL_TYPES) if v == self.animal_type)
+
+        self.competition_score = (type_score + 1) * (quality_score + 1)  # Should be positive in all cases
+    
+    def get_weight_adjusted_competition_score(self):
+        """Potential to add weight as a multiplier function.
+
+           E.g. weight the competition score of
+           each animal by predator_weight/ prey_weight so that heavier prey have their score weighted
+           down and are therefore less likely to be eaten. But still occupy the same location in the food
+           chain when compared to larger predators.
+
+           For example:
+                ELEPHANT_WEIGHT = 1000 Kg
+                LION_WEIGHT = 20 Kg
+                WEIGHT_MULTIPLIER = 20/1000 * CONSTANT 10
+                ADJUSTED_COMPETITION_SCORE = 20 percent of elephants original competitiveness
+                """
+        raise NotImplementedError
+
+    def get_prey_relationships(self):
+        """Undefined method designed to be overridden.
+
+        Potentially - Identify where an animal sits in a food web based on
+        it's species/size/weight... Generate a score to represent it's
+        location.
+        """
+        raise NotImplementedError
+
+
+class Lion(BaseSpecies):
```

Comment:
See if you can think of a way of configuring these subclasses without extending any methods (just setting attributes on the class).